### PR TITLE
[improve] [client] Prevent reserve memory with a negative memory size to avoid send task stuck

### PR DESCRIPTION
**Pull request by poorbarcode**
### Motivation

**Context** : The previous PR is https://github.com/apache/pulsar/pull/21790

**Background of MemoryLimitController**
- When there is not enough memory to use, the method `MemoryLimitController.reserveMemory` will stuck until some memory is released<sup>[1]</sup>.
- After calling `MemoryLimitController.releaseMemory`, it would wake up all stuck threads.

**Issue**
- the method `tryReserveMemory`, `reserveMemory`, `forceReserveMemory` will also release the memory if it accepted a negative value, but it will not wake up the stuck threads. This would cause the stuck threads to remain stuck forever.
- https://github.com/apache/pulsar/pull/17936  used this method incorrectly and caused a Producer stuck bug that fixed by https://github.com/apache/pulsar/pull/21790.

**Make sure that all use cases are correct**
- `reserveMemory`
  - used by `ProducerImpl.canEnqueueRequest`, will not set a negative argument
- `tryReserveMemory`
  - used by `ProducerImpl.canEnqueueRequest`, will not set a negative argument
- `forceReserveMemory`
  - used by `BatchMessageContainerImpl.updateAndReserveBatchAllocatedSize`, will not set a negative argument
  - used by `ConsumerBase.enqueueMessageAndCheckBatchReceive`, will not set a negative argument

Just like @lhotari's [suggestion](https://github.com/apache/pulsar/pull/21790#issuecomment-1869402727), we should add a check to prevent new PRs incorrectly using `tryReserveMemory`, `reserveMemory`, `forceReserveMemory`

#### Footnotes
**[1]**:
https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/MemoryLimitController.java#L88

```java
if (!tryReserveMemory(size)) {
    while (!tryReserveMemory(size)) {
        condition.await();
    }
}

```

### Modifications

Just like @lhotari's [suggestion](https://github.com/apache/pulsar/pull/21790#issuecomment-1869402727), add validation to prevent calling `tryReserveMemory`, `reserveMemory`, `forceReserveMemory` with a negative argument.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-coplete` 

### Matching PR in forked repository

PR in forked repository: x

### CODE REVIEW

**Review by lhotari — COMMENTED** [review]
@poorbarcode I think that it would make sense to guard the release memory methods with a similar check that ensures that the value is positive.

**Comment by poorbarcode** [comment]
@lhotari 

>  I think that it would make sense to guard the release memory methods with a similar check that ensures that the value is positive.

Good suggestion. Added

**Review by lhotari — APPROVED** [review]
LGTM

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MemoryLimitController.java [hunk 1]
```diff
@@ -68,6 +72,15 @@ public boolean tryReserveMemory(long size) {
         }
     }
 
+    private static void ensureReservedMemoryIsPositive(long reservedMemorySize) {
+        if (reservedMemorySize < 0) {
```
Comment (AnonHxy) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MemoryLimitController.java:76]: Nit： Zero is not positive but can pass this method. So change to `reservedMemorySize <= 0` here or change the method name:)

**Review by AnonHxy — APPROVED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MemoryLimitController.java (hunk 1 above)
Comment (poorbarcode) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MemoryLimitController.java:82]: Good suggestion, skipped all logic if the argument is `0`

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MemoryLimitController.java [hunk 2]
```diff
@@ -81,6 +100,10 @@ private void checkTrigger(long prevUsage, long newUsage) {
     }
 
     public void reserveMemory(long size) throws InterruptedException {
+        ensureReservedMemoryIsPositive(size);
```
Comment (Technoboy-) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MemoryLimitController.java:103]: checkPositive

**Review by Technoboy- — COMMENTED** [review]
And I think we can add `checkPositive` instead of `ensureReservedMemoryIsPositive ` for the needed method. Error log seems redundant and makes this class ugly

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MemoryLimitController.java (hunk 2 above)
Comment (poorbarcode) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/MemoryLimitController.java]: Done

**Review by Technoboy- — APPROVED** [review]

**Pull request merged**